README.md
# Sketches

* [github.com/postmodern/sketches/](http://github.com/postmodern/sketches/)
* [github.com/postmodern/sketches/issues/](http://github.com/postmodern/sketches/issues/)
* Postmodern (postmodern.mod3 at gmail.com)

## Description

Sketches allows you to create and edit Ruby code from the comfort of your
editor, while having it safely reloaded in IRB whenever changes to the
code are saved.

## Features

* Spawn an editor of your choosing from IRB.
* Automatically reload your code when it changes.
* Run the editor in the background.
* Run the editor in a new terminal.
* Use a custom editor command.
* Use a custom terminal command.
* Use a custom temp directory to store sketches in.

## Problems

* Local variables cannot be loaded. Ruby defines local variables as local
  to the file which they were originally defined in.
* No support for Inotify or FSevent _yet_. It would be nice to attempt to
  use Inotify or FSevent bindings, if available, for efficient polling of
  sketch files.

## Install

    $ sudo gem install sketches

Then require sketches in your <tt>.irbrc</tt> file:

    require 'sketches'

Sketches can be configured to use a custom editor command:

    Sketches.config :editor => 'gvim'

Sketches can also be configured to run in a new terminal:

    Sketches.config :editor => 'vim',
                    :background => true,
                    :terminal => lambda { |cmd|
                      "xterm -fg gray -bg black -e #{cmd.dump}"
                    }

## Examples

Open a new sketch:

    sketch

Open a new named sketch:

    sketch :foo

Open a sketch from an existing file:

    sketch_from 'path/to/bar.rb'

Reopen an existing sketch:

    sketch 2

    sketch :foo

List all sketches:

    sketches

Name a sketch:

    name_sketch 2, :foo

Save a sketch to an alternant location:

    save_sketch :foo, 'path/to/foo.rb'

## License

Sketches - A Ruby library for live programming and code reloading.

Copyright (c) 2009 Hal Brodigan (postmodern.mod3 at gmail.com)

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA


History.txt
=== 0.1.1 / 2009-12-12

* Added Config.terminal.
* Added Config.terminal=.
* Added Config.background.
* Added Config.background=.
* Allow the editor command to be ran in a new terminal.
* Allow running the editor command in the background.
* Allow sketch files to be evaled immediately after exiting the editor
  (thanks Jonathan Penn).
* Fixed a bug with TempSketch.open on JRuby (thanks Sergey Moshnikov).

=== 0.1.0 / 2009-04-29

* Initial release:
  * Spawn an editor of your choosing from IRB.
  * Automatically reload your code when it changes.
  * Use a custom editor command.
  * Use a custom temp directory to store sketches in.



lib/sketches/sketches.rb
#
#--
# Sketches - A Ruby library for live programming and code reloading.
#
# Copyright (c) 2009 Hal Brodigan (postmodern.mod3 at gmail.com)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
#Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#++
#

require 'sketches/config'
require 'sketches/cache'

module Sketches
  #
  # Configure Sketches with the given _options_.
  #
  # _options_ may contain the following keys:
  # <tt>:tmpdir</tt>:: Directory to store temporary sketches in.
  #                    Defaults to Dir.tmpdir if unspecified.
  # <tt>:background</tt>:: Specifies wether to run the editor as a
  #                        background or a foreground process.
  # <tt>:terminal</tt>:: The terminal to optionally run the editor within.
  # <tt>:editor</tt>:: The editor to use to edit sketches.
  # <tt>:pause</tt>:: The number of seconds to pause in-between
  #                   checking if any sketches were modified.
  #
  #   Sketches.config :editor => 'gvim', :pause => 2
  #
  #   Sketches.config :editor => 'vim',
  #                   :background => true,
  #                   :terminal => lambda { |cmd|
  #                     "xterm -fg gray -bg black -e #{cmd.dump}"
  #                   }
  #
  def Sketches.config(options={})
    if options[:tmpdir]
      Config.tmpdir = options[:tmpdir]
    end

    if options[:background]
      Config.background = options[:background]
    end

    if options[:terminal]
      Config.terminal = options[:terminal]
    end

    if options[:editor]
      Config.editor = options[:editor]
    end

    if options[:pause]
      Config.pause = options[:pause]
    end

    return nil
  end

  @@sketches_cache = nil

  #
  # The cache of sketches.
  #
  def Sketches.cache
    unless @@sketches_cache
      @@sketches_cache = Cache.new
    end

    return @@sketches_cache
  end

  #
  # Edits the sketch with the specified _id_or_name_. If no sketch exists
  # with the specified _id_or_name_, one will be created.
  #
  #   Sketches.sketch 2
  #
  #   Sketches.sketch :foo
  #
  def Sketches.sketch(id_or_name)
    Sketches.cache.synchronize do
      sketch = Sketches.cache[id_or_name]
      sketch ||= Sketches.cache.new_sketch(id_or_name)

      sketch.synchronize { sketch.edit }
    end
  end

  #
  # Creates a new sketch using the specified _path_.
  #
  #   Sketches.from 'path/to/foo.rb'
  #
  def Sketches.from(path)
    Sketches.cache.synchronize do
      sketch = Sketches.cache.reuse_sketch(path)

      sketch.synchronize { sketch.edit }
    end
  end

  #
  # Names the sketch with the specified _id_ with the specified _name_.
  #
  #   Sketches.name 2, :foo
  #
  def Sketches.name(id,name)
    Sketches.cache.synchronize do
      Sketches.cache.name_sketch(id,name)
    end
  end

  #
  # Saves the sketch with the specified _id_or_name_ to the specified
  # _path_.
  #
  #   Sketches.save 2, 'path/to/example.rb'
  #
  #   Sketches.save :foo, 'path/to/foo.rb'
  #
  def Sketches.save(id_or_name,path)
    Sketches.cache.synchronize do
      if (sketch = Sketches.cache[id_or_name])
        sketch.save(path)
      end
    end
  end

  #
  # Print out all of the sketches.
  #
  def Sketches.print
    Sketches.cache.synchronize { puts Sketches.cache }
  end
end
